Add MetrLA Dataset

radandreicristian opened the pull request:
With this pull request, the [MetrLA](https://paperswithcode.com/sota/traffic-prediction-on-metr-la) dataset and helper classes were added, including the following files:

- Dataset class
- I/o methods
- Unit tests
- Documentation

radandreicristian committed:
Add MetrLa dataset
Add MetrLa i/o utils
Add MetrLa unit test

radandreicristian committed:
Solve indexing problem with data saving
Fix issues as per unit tests

radandreicristian committed:
Update unit test
Fix a bug with NaN edge attributes
Update docstrings to match Google format

radandreicristian committed:
Remove unnecessary modification to dataset.py

radandreicristian committed:
Update docstrings

radandreicristian committed:
Update docstrings (2)

radandreicristian committed:
Update unit test

radandreicristian committed:
Add warning for long unit test

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

radandreicristian committed:
Merge branch 'master' into dataset_metrla

radandreicristian commented:
Well, this is awkward. The docs/unit tests are all working locally (Ubuntu 21.04, Python 3.9, Torch 1.10.1+CU11.3). Any input on how to deal with this would be greatly appreciated, as this is only my second contribution to open-source:)

rusty1s committed:
Merge branch 'master' into dataset_metrla

rusty1s commented:
I think I fixed this in master, see https://github.com/pyg-team/pytorch_geometric/commit/e04fe09c6937e37d4dd4bb9923d1ed5dabf4bc9c. You might need to rebase your fork from master.

rusty1s commented:
Replacing all occurrences of `list[...]` with `List[...]` should resolve the remaining errors.

radandreicristian committed:
Merge remote-tracking branch 'upstream/master' into dataset_metrla

radandreicristian committed:
Merge remote-tracking branch 'origin/dataset_metrla' into dataset_metrla

radandreicristian commented:
> Replacing all occurrences of `list[...]` with `List[...]` should resolve the remaining errors.

Thanks for the help, working on it!

radandreicristian committed:
Update type hinting from list[...] to typing.List[...] for py<3.9 compat

radandreicristian commented:
There's another error requiring "tables" (for pandas reading from HDF files). I'll add it to the req's on my branch and try again.

rusty1s commented:
Any chance to get rid of this additional dependency? Otherwise, we may want to add it to the `full_requires` dependency list.

radandreicristian committed:
Add missing requirement to setup.py for HD5 parsing

radandreicristian commented:
Yes, perhaps I can convert the original .hd5 file to .csv.

rusty1s commented on setup.py:
```diff
@@ -4,17 +4,9 @@
-    'googledrivedownloader',
+    'tqdm', 'yacs', 'numpy', 'scipy', 'pandas', 'jinja2', 'requests',
+    'pyparsing', 'hydra-core', 'scikit-learn', 'googledrivedownloader',
+    'tables'
```
```suggestion
    'tables',
```

rusty1s commented on test/datasets/test_metrla.py:
```diff
@@ -0,0 +1,38 @@
+
+
+def test_metrla() -> None:
+    root = '/tmp/metr_la'
```
```suggestion
    root = osp.join('/tmp', 'metr_la')
```
As the tests suggests, this might need to be fixed on Windows, *e.g.*, via `osp.join('/tmp', 'metr_la')`.

rusty1s commented on test/datasets/test_metrla.py:
```diff
@@ -0,0 +1,38 @@
+    n_previous_steps = 6
+    n_future_steps = 6
+
+    print("This may take a couple of minutes...")
```
How long does it take to download the dataset? We generally do not test downloading and processing datasets as part of our CI. I'm fine with dropping this test completely.

radandreicristian replied:
Downloading takes like 10 sec, processing takes a good 2-3 minutes. I can try to limit it to generate a single example and do the last 5 asserts on it, otherwise will just drop it.

rusty1s commented on torch_geometric/datasets/__init__.py:
```diff
@@ -157,6 +158,7 @@
     'RandomPartitionGraphDataset',
     'LINKXDataset',
     'EllipticBitcoinDataset',
+    'MetrLa'
```
```suggestion
    'MetrLa',
```

rusty1s commented on torch_geometric/datasets/metrla.py:
```diff
@@ -0,0 +1,177 @@
+class MetrLa(Dataset):
+    r"""The Los Angeles Metropolitan (MetrLA) highway traffic dataset
+    introducted in `"Big Data and its Technical Challenges"
+    <http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.681.7248&rep=rep1&type=pdf>`_
```
We may need to split the URL into 2 lines:
```
    <http://citeseerx.ist.psu.edu/viewdoc/download?
    doi=10.1.1.681.7248&rep=rep1&type=pdf>`_
```

rusty1s commented on torch_geometric/datasets/metrla.py:
```diff
@@ -0,0 +1,177 @@
+
+class MetrLa(Dataset):
+    r"""The Los Angeles Metropolitan (MetrLA) highway traffic dataset
+    introducted in `"Big Data and its Technical Challenges"
```
```suggestion
    introducted in the `"Big Data and its Technical Challenges"
```

rusty1s commented on torch_geometric/datasets/metrla.py:
```diff
@@ -0,0 +1,177 @@
+    volumes (how many cars have passed by a sensor), aggregated in 5 minutes
+    intervals. The traffic prediciton task is a time series regression,
+    which aims to predict the traffic volumen values in each node for the
+    next *n_next* time steps, given the volumnes in the *n_prev*  previous
```
```suggestion
    next :ob:`n_next` time steps, given the volumes in the :obj:`n_prev`  previous
```

rusty1s commented on torch_geometric/datasets/metrla.py:
```diff
@@ -0,0 +1,177 @@
+    intervals. The traffic prediciton task is a time series regression,
+    which aims to predict the traffic volumen values in each node for the
+    next *n_next* time steps, given the volumnes in the *n_prev*  previous
+    time steps. Both n_next and n_prev are pre-set values. Given the
```
```suggestion
    time steps. Both :obj:`n_next` and :obj:`n_prev` are pre-set values. Given the
```

rusty1s commented on torch_geometric/datasets/metrla.py:
```diff
@@ -0,0 +1,177 @@
+    The initial adjacency matrix is constructed by applying a thresholded
+    Gaussian kernel over the real-world distances between the sensors:
+
+    :math:`W_{ij} = exp(-\frac{dist(v_{i}, v_{j})^2}{\sigma^{2}})`,
```
Let's put that formula in equation mode:
```
.. math::
```

rusty1s commented on torch_geometric/datasets/metrla.py:
```diff
@@ -0,0 +1,177 @@
+    pre-fixed threshold, and dist(, ) is the real-world road distance
+    between two sensors, and :math:`\sigma^{2}` is the variance of the
+    distances between sensors. A zero-value in the adjacency matrix does
+    not indicate that thers is no road connection between two sensors (
```
```suggestion
    not indicate that there is no road connection between two sensors (
```

rusty1s commented on torch_geometric/datasets/metrla.py:
```diff
@@ -0,0 +1,177 @@
+        after the first time, so that's not reliable. Open to any
+        workarounds, but the dataset has a fixed size, so
+        this is not a deal breaker imo."""
+        self.n_readings = 34272
```
It's probably sufficient to just test that the first few processed files exists. Otherwise, checking the existence of 34272 can quickly become the bottleneck.

radandreicristian replied:
Checking the existence of ~34K file takes 1.2 seconds in my case (on SSD; HDD storage might be slower indeed), so it's probably not that much of a bottleneck for something that happens every time the code is run (if I remove the test case).

Anyway, your suggestion is valid, and we should optimize wherever possible. I modified it to check only the first 5 (needed a bit more changes in the code than replacing the value with 5, i.e. in the "len" method), and it's coming with the next push. Thanks!

rusty1s commented on torch_geometric/datasets/metrla.py:
```diff
@@ -0,0 +1,177 @@
+    def processed_file_names(self) -> str:
+        r"""The name of the files in the :obj:`self.processed_dir`
+        folder that must be present in order to skip processing."""
+        return [f'data_{i}.pt' for i in range(self.io.dataset_len)
```
```suggestion
        return [f'data_{i}.pt' for i in range(5)
```

radandreicristian replied:
Right, but then the problem is in the "get" function. But I guess I can solve that with an f-string (assuming all the files are named with the same template). Is that an appropriate way to go?

radandreicristian replied:
I wrapped the getter of the name in a private method (as it was used more than twice, to avoid further problems with renaming, and to avoid possible typos with hard-coded values throughout the code).

rusty1s commented on torch_geometric/datasets/metrla.py:
```diff
@@ -0,0 +1,177 @@
+        for idx in range(self.io.dataset_len):
+            # Select the "slice" among the first dimension
+            # x_index (n_previous_steps, n_nodes, n_features)
+            x_index = torch.tensor(data=x[idx, ...], dtype=torch.float32)
```
```suggestion
            x_index = x[idx]
```
Isn't that sufficient?

radandreicristian replied:
That's just to convert from numpy to torch tensor, since the Data constructor specifies x to be Optional[Tensor].

rusty1s commented on torch_geometric/datasets/metrla.py:
```diff
@@ -0,0 +1,177 @@
+                                     self.edge_attributes_file_name)
+
+        data = torch.load(f=Path(self.processed_dir) / file_name)
+        setattr(data, 'edge_index', edge_index)
```
```suggestion
        data.edge_index = edge_index
```

radandreicristian committed:
Replace HDF file with CSV to avoid extra dependency
Update io loading to CSV
Remove tables from requirements

radandreicristian commented on torch_geometric/datasets/metrla.py:
```diff
@@ -0,0 +1,177 @@
+                ] + ["edge_index.pt", "edge_attributes.pt"]
+
+    @property
+    def edge_index_file_name(self) -> str:
```
They're not necessary, but since their values are used more than 2 times throughout the code I figured it would be more elegant this way. Would definitely make sense to have them private, thanks for pointing out!

radandreicristian commented on torch_geometric/datasets/metrla.py:
```diff
@@ -0,0 +1,177 @@
+
+    Args:
+        root (string): Root directory where the dataset should besaved.
+        n_previous_steps (int): The number of previous time steps to consider
```
Right, I'll just update them in the docstring instead of changing the code.

radandreicristian committed:
Remove bottleneck unit test
Fix typos and mistakes in docstrings
Speed up verification of processed files by limiting the range to 5
Replace setattr calls with direct assignation

radandreicristian committed:
Fix flake8 error

radandreicristian committed:
Fix sphinx warning/error

radandreicristian commented:
Alright, integrated all the feedback and pushed. 

Thanks for the guidance, @rusty1s!

radandreicristian committed:
Merge branch 'master' into dataset_metrla

radandreicristian committed:
Remove a leftover print

radandreicristian committed:
Merge branch 'dataset_metrla' of https://github.com/radandreicristian/pytorch_geometric into dataset_metrla
Merge latest changes from upstream/master

radandreicristian committed:
Add method to retrieve adjacency matrix without accessing an individual data point

radandreicristian committed:
Add in memory version of the dataset

radandreicristian committed:
Merge branch 'master' into dataset_metrla

radandreicristian committed:
Automatically return day-of-week and hour-of-day features.

radandreicristian committed:
Merge branch 'dataset_metrla' of https://github.com/radandreicristian/pytorch_geometric into dataset_metrla

radandreicristian committed:
Remove normalization

radandreicristian committed:
Update temporal features indexing.

radandreicristian committed:
Update temporal features.
Clear code.
Rename variables.
